### PR TITLE
feat: duplicate email prevention and email/name normalization

### LINKED ISSUE #120: Unable to create user in /admin
Image

I am trying to create the user above, but when I click on "Save changes" nothing happens.

### PR DESCRIPTION
**Pull request by angelospk**
Closes #120

## Description

**Summary:**
This Pull Request improves the reliability, security, and developer experience of the user creation and management flow. It adds comprehensive input normalization, prevents silent failures on duplicate user creation, and ensures cleaner error feedback and database state.

**Changes Made:**

* **Data Layer & Validation ([`src/lib/db/users.ts`](./src/lib/db/users.ts))**:
  - **Input Normalization**: Added `normalizeEmail`, `normalizeName`, and `normalizeAdminUserData` functions. These automatically trim extra whitespaces, convert emails to lowercase, and handle empty names before they reach the database layer.
  - **Duplicate Email Prevention**: Added proactive checks in `createUser` and `updateUser` to ensure no conflicting emails exist.
  - **Clearer Error Surfacing**: Handled both custom checks and Prisma duplicate constraint (`P2002`) errors to consistently throw: `"A user with this email already exists."`. This fixes the root cause of the silent UI failures.

* **Utilities ([`src/lib/utils.ts`](./src/lib/utils.ts))**:
  - **Whitespace Safety**: Hardened the `klitiki` name formatting function to correctly parse, trim, and evaluate strings containing extra or trailing spaces.

* **API Routes ([`src/app/api/admin/users/route.ts`](./src/app/api/admin/users/route.ts))**:
  - **Cleanup of Orphan Records**: The invite verification token is now immediately proactively cleared if the email sending step fails.
  - **Improved Logging Security**: Removed Personally Identifiable Information (PII) from invite failure logs. The logs now safely rely solely on `user.id` instead of the email.

* **Frontend UI ([`src/components/admin/users/user-dialog.tsx`](./src/components/admin/users/user-dialog.tsx))**:
  - **Refactored Validation**: Removed redundant client-side empty-email checks, centralizing validation.
  - **Enhanced Error Feedback**: Added runtime-safe parsing for server validation errors so that the Duplicate Email message actually renders properly instead of displaying `[object Object]`.

* **Testing ([`src/lib/__tests__/users.test.ts`](./src/lib/__tests__/users.test.ts), [`src/app/api/admin/users/route.test.ts`](./src/app/api/admin/users/route.test.ts), [`src/lib/__tests__/utils.test.ts`](./src/lib/__tests__/utils.test.ts))**:
  - Extensive tests added to assert new behavior for `updateUser`/`createUser` normalization, constraints, duplicate email errors, `klitiki` whitespaces, and route cleanup edge cases.

**Testing:**
- Verified visually that the "Duplicate Email" error now properly surfaces in the UI.
- All new and existing test suites pass under the current test setup (`Next.js` and `db`).

<h3>Greptile Summary</h3>

This PR adds input normalization (email trimming/lowercasing, name trimming), duplicate-email detection via P2002 error mapping, and a refactored invite flow in `src/lib/auth/invite.ts` with proper try/catch and token cleanup. Most previously-flagged concerns — `generateSignInLink` inside try/catch, Discord alert placement, DELETE response format mismatch, and P2025 handling — have been resolved in fixup commits, leaving only minor style-level observations.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all previously-raised P1 issues have been addressed and no new P0/P1 issues were found

All remaining findings are P2 or lower. The normalization logic, error mapping, invite flow, and token cleanup are all correctly implemented with solid test coverage. The fixup commits resolved the critical concerns from the prior review round.

No files require special attention

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/lib/auth/invite.ts | New module extracts invite email sending with proper try/catch and token cleanup on failure |
| src/lib/db/users.ts | Adds email/name normalization and maps P2002/P2025 Prisma errors to typed ApiError subclasses |
| src/app/api/admin/users/route.ts | Cleaned up to use handleApiError and sendInviteEmail; returns warning JSON on partial-success invite failure |
| src/components/admin/users/user-dialog.tsx | Parses JSON error responses and shows warning toast when user is created but invite email fails |
| src/lib/utils.ts | Hardens klitiki to handle whitespace-only strings, hyphenated names, and multi-word Greek names |
| src/lib/__tests__/users.test.ts | Comprehensive tests for normalization, P2002/P2025 mapping, and both direct and transaction updateUser paths |
| src/app/[locale]/(other)/admin/users/page.tsx | Fixed DELETE error parsing from response.text() to response.json() to match handleApiError JSON format |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Admin as Admin UI
    participant Route as POST /api/admin/users
    participant DB as users.ts (createUser)
    participant Invite as invite.ts (sendInviteEmail)
    participant Email as Email Provider

    Admin->>Route: POST {email, name, ...}
    Route->>DB: createUser(data)
    DB->>DB: normalizeAdminUserData()
    alt Duplicate email
        DB-->>Route: ConflictError (409)
        Route-->>Admin: 409 {error}
    else Valid data
        DB-->>Route: newUser
        Route->>Invite: sendInviteEmail(email, name)
        Invite->>Invite: generateSignInLink() → verificationToken
        Invite->>Email: sendEmail(html)
        alt Email sent OK
            Email-->>Invite: success
            Invite-->>Route: true
            Route->>Route: sendUserOnboardedAdminAlert()
            Route-->>Admin: 200 {user}
        else Email failed
            Email-->>Invite: failure
            Invite->>Invite: deleteMany(verificationToken)
            Invite-->>Route: false
            Route-->>Admin: 200 {user, warning}
        end
    end
```

<sub>Reviews (23): Last reviewed commit: ["feat: duplicate email prevention with DB..."](https://github.com/schemalabz/opencouncil/commit/18256a4ef550ccdc252a8a1b3821df6a9834e574) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=23193466)</sub>

### CODE REVIEW

File: src/app/api/admin/users/route.ts [hunk 1]
```diff
@@ -8,8 +8,30 @@ import { createHash } from "crypto"
...
+import { PrismaError } from "@/lib/db/types"
 
-async function generateSignInLink(email: string) {
+
+interface VerificationTokenKey {
+    identifier: string
+    token: string
+}
+
+function getUserErrorResponse(error: unknown, fallbackMessage: string): NextResponse {
```
Comment (kouloumos) [inline src/app/api/admin/users/route.ts]: This reimplements what `handleApiError` in `src/lib/api/errors.ts` already does. Have the db layer throw `ConflictError("A user with this email already exists.")` and `BadRequestError(...)` instead, then use `handleApiError(error, "Failed to create user")` here. That way the status code lives with the error, not in a string-matching function.

File: src/lib/db/users.ts [hunk 2]
```diff
@@ -67,11 +99,27 @@ export async function createUser(data: AdminUserData, options: { skipAuthCheck?:
...
+                equals: email,
+                mode: "insensitive",
+            },
+        },
+        select: { id: true },
+    });
+
+    if (existingUser) {
+        throw new Error("A user with this email already exists.");
+    }
```
Comment (kouloumos) [inline src/lib/db/users.ts]: `email` already has a `@unique` constraint in the schema. The P2002 catch below handles duplicates correctly. This `findFirst` adds an extra query on every create/update and still has a TOCTOU race condition. Drop it and rely on the P2002 catch -- that's the one that's actually atomic.

**Comment by kouloumos** [comment]
I force-pushed to restructure the branch from 12 commits into 3 clean logical commits:

1. `fix: improve klitiki utility for hyphenated names and edge cases`
2. `refactor: extract invite email logic to shared module`
3. `feat: duplicate email prevention with DB-layer normalization and error mapping`

The previous branch tip is preserved as `backup/pr-221-pre-restructure`. You can verify no code was lost:

```bash
git diff backup/pr-221-pre-restructure fix/add-validation-to-prevent-duplicate -- src/
# Expected: empty output (identical source code)
```
